### ISSUE TITLE
we don't have release store?

### ISSUE THREAD

**Issue by crusader-mike** [jfbastien/no-sane-compiler#4]
Hey, JF

Watched your presentation on youtube and it raised few questions. I don't use twitter, so an issue on a github seems like a good way to contact you :-)

1. We don't have "release store"? Did you mean "we don't have release **load**"?

2. Example with progress bar doesn't make any sense to me -- why would compiler optimize code by fusing atomic ++progress operations?

    ++progress is the same as fetch_add(1, **std::memory_order_seq_cst**)+1

where memory model is described as:

> Any operation with this memory order is both an acquire operation and a release operation, plus a single total order exists in which all threads observe all modifications in the same order

So, basically, if you unroll the loop -- not a single relaxed operation can cross the boundaries defined by ++progress statements, nor these statements can reorder wrt each other (or even get any closer to each other).

Can you elaborate why you believe that fusion is allowed here?

3. Same but simpler case -- fusing together two atomic increments. They both use seq_cst model -- why do you believe it is ok for compiler to fuse them?

Regards,
Michael.

**Comment by jfbastien** [comment]
1. Yeah, release load 😊 Right after the talk Fedor pointed out I'd mis-spoken.
2. Hans and I discuss this example [here](http://wg21.link/p0062).
3. Same as above: there's no synchronization with anyone! It's totally possible to have the "observer" thread not notice any progress because it doesn't get schedule until work is done. In that execution, the outcome is indistinguishable from the proposed transformation... so how can you tell which occurred?

Not to say a compiler *should* do this! The entire point of my paper was to point out what the standard allowed. Some good, some bad!

**Comment by crusader-mike** [comment]
1. Well, you also mis-typed -- that slide was saying "store" ;)
2. I read the P0062R1. Not really sure what is the takeaway from all this -- are you saying that current standard should allow optimization of atomics? Or that it already allows and it should be clamped down?

 If later -- why do you believe that fusing atomics is allowed by standard? I.e. which provisions allow that?

3. Yes, it is totally possible for observer to miss few updates to atomic variable -- why it means it is ok to fuse them? What if one op is "inc" and another is "dec" -- you can't justify fusing these just because observers can miss some of them.

The way I look at this -- memory model defines order of things and there is a total ordering between release writes to same variable. Which means that first write should **happen** before the second. It may or may not be observable by other thread, but it is beside the point. Really curious why you believe standard is ok with fusing...

**Comment by jfbastien** [comment]
The feeling in the concurrency and parallelism group was that something's underspecified, but the proposal in P0062R1 wasn't right, and neither was adding a note discouraging compilers (because notes are non-normative and don't really mean much).

Optimizations are allowed by the current standard. Some are good, some aren't.

> Yes, it is totally possible for observer to miss few updates to atomic variable -- why it means it is ok to fuse them? What if one op is "inc" and another is "dec" -- you can't justify fusing these just because observers can miss some of them.

You can, and the standard does. In fact, I've worked on machines whose hardware does exactly this and where these operations would usually be invisible (barring page faults and such). Is that hardware wrong?

To try to give this a more intuitive explanation: this specific optimization can be done under C++'s as-if rule: it's as if the execution occurs with the race unobserved, because the only way a program could be written to guaranteed observing the race is if it's racy which makes it UB, and you can assume there's no UB.

The standard walks a fine line between what it thinks the langue does, what compilers should do, and what actual hardware does. It's quite hard to legislate existing hardware.

At the end of the day the goal of concurrency and parallelism in the language is providing easy to use maximal performance with minimal surprise.

**Comment by crusader-mike** [comment]
> Is that hardware wrong?

Well, it isn't about hardware -- C++ operates within it's memory model. As a developer I care not about underlying hardware -- I write program for that memory model and rely on compiler to implement it on given hardware properly. It is up to language designers to pick such model that can be efficiently mapped to most hardware out there.

But I see your point. You are saying that in case of two updates to same variable C++ memory model does not guarantee that observer can see effect of first one. Can you point me to specific section in standard maybe? (the one that describes memory model)

**Comment by jfbastien** [comment]
P0062R1 and references document the "why?" in the standard. It never says "you can do this". It doesn't say otherwise either.
